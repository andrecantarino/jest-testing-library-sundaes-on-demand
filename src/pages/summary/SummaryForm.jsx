import React, { useState } from "react";
import Form from "react-bootstrap/Form";
import FormCheck from "react-bootstrap/FormCheck";
import FormGroup from "react-bootstrap/FormGroup";
import Button from "react-bootstrap/Button";

const SummaryForm = () => {
  const [tcChecked, setTcChecked] = useState(false);
  const checkboxLabel = (
    <span>
      I agree to <span style={{ color: "blue" }}>Terms and conditions</span>
    </span>
  );

  return (
    <Form>
      <FormGroup controlId="terms-and-conditions">
        <FormCheck
          type="checkbox"
          checked={tcChecked}
          onChange={(e) => setTcChecked(e.target.checked)}
          label={checkboxLabel}
          role="checkbox"
        />
      </FormGroup>
      <Button
        variant="primary"
        type="submit"
        role="button"
        disabled={!tcChecked}
      >
        Confirm order
      </Button>
    </Form>
  );
};

export default SummaryForm;
