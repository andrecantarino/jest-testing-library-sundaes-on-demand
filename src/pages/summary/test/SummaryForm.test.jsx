import {
  render,
  screen,
  waitForElementToBeRemoved,
} from "@testing-library/react";
import SummaryForm from "../SummaryForm";
import userEvent from "@testing-library/user-event";

it("should get a snapshot", () => {
  const { asFragment } = render(<SummaryForm />);

  expect(asFragment(<SummaryForm />)).toMatchSnapshot();
});

test("Initial conditions", () => {
  render(<SummaryForm />);
  const checkbox = screen.getByRole("checkbox", {
    name: /I agree to Terms and conditions/i,
  });
  expect(checkbox).not.toBeChecked();

  const confirmButton = screen.getByRole("button", { name: /confirm order/i });
  expect(confirmButton).toBeDisabled();
});

test("Checkbox enables button on first click and disables on second click", () => {
  render(<SummaryForm />);
  const checkbox = screen.getByRole("checkbox", {
    name: /I agree to Terms and conditions/i,
  });
  const confirmButton = screen.getByRole("button", { name: /confirm order/i });

  userEvent.click(checkbox);
  expect(checkbox).toBeChecked();

  userEvent.click(checkbox);
  expect(confirmButton).toBeDisabled();
});

test("popover responds to hover", async () => {
  render(<SummaryForm />);

  const nullPopover = screen.queryByText(
    /no ice cream will actually be delivered/i
  );
  expect(nullPopover).not.toBeInTheDocument();

  const termsAndConditions = screen.getByText(/terms and conditions/i);
  userEvent.hover(termsAndConditions);

  const popover = screen.getByText(/no ice cream will actually be delivered/i);
  expect(popover).toBeInTheDocument();

  userEvent.unhover(termsAndConditions);
  await waitForElementToBeRemoved(() =>
    screen.queryByText(/no ice cream will actually be delivered/i)
  );
});
