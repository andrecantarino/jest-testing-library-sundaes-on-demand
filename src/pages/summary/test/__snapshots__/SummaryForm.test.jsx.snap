// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should get a snapshot 1`] = `
<DocumentFragment>
  <form
    class=""
  >
    <div>
      <div
        class="form-check"
      >
        <input
          class="form-check-input"
          id="terms-and-conditions"
          role="checkbox"
          type="checkbox"
        />
        <label
          class="form-check-label"
          for="terms-and-conditions"
          title=""
        >
          <span>
            I agree to
            <span
              style="color: blue;"
            >
              Terms and conditions
            </span>
          </span>
        </label>
      </div>
    </div>
    <button
      class="btn btn-primary"
      disabled=""
      role="button"
      type="submit"
    >
      Confirm order
    </button>
  </form>
</DocumentFragment>
`;
